bio2zarr/bed2zarr.py

```diff
@@ -0,0 +1,559 @@
+            chunks=[metadata.num_contigs],
+            dimensions=["contigs"],
+            description="Contig ID",
+        )
+        name_id_spec = ZarrArraySpec.new(
+            name="name_id",
+            dtype="O",
+            shape=[metadata.num_names],
+            chunks=[metadata.num_names],
+            dimensions=["names"],
+            description="Name ID",
+        )
+        specs.extend([contig_id_spec, name_id_spec])
+        return BedZarrSchema(
+            format_version=ZARR_SCHEMA_FORMAT_VERSION,
+            fields=specs,
+            contigs=metadata.contigs,
+            names=metadata.names,
+            bed_type=metadata.bed_type.name,
+            records_chunk_size=records_chunk_size,
+        )
+
+
+@dataclasses.dataclass
+class BedFieldSummary(core.JsonDataclass):
+    num_chunks: int = 0
+    compressed_size: int = 0
+    uncompressed_size: int = 0
+    max_value: Any = -math.inf
+    min_value: Any = math.inf
+
+    def update(self, other):
+        self.num_chunks = other.num_chunks
+        self.compressed_size = other.compressed_size
+        self.uncompressed_size = other.uncompressed_size
+        self.min_value = min(self.min_value, other.min_value)
+        self.max_value = max(self.max_value, other.max_value)
+
+    @staticmethod
+    def fromdict(d):
+        return BedFieldSummary(**d)
+
+
+@dataclasses.dataclass
+class BedField:
+    category: str
+    name: str
+    alias: str
+    description: str
+    bed_dtype: str
+    summary: BedFieldSummary
+
+    def smallest_dtype(self):
+        """Return the smallest dtype suitable for this field based on
+        type and values"""
+        s = self.summary
+        if self.bed_dtype == "Integer":
+            if not math.isfinite(s.max_value):
+                ret = "i1"
+            else:
+                ret = core.min_int_dtype(s.min_value, s.max_value)
+        elif self.bed_dtype == "Character":
+            ret = "U1"
+        elif self.bed_dtype == "Category":
+            ret = core.min_int_dtype(s.min_value, s.max_value)
+        else:
+            assert self.bed_dtype == "String"
+            ret = "O"
+        return ret
+
+
+def guess_bed_file_type(path):
+    """Check the number of columns in a BED file and guess BED
+    type."""
+    first = pd.read_table(path, header=None, nrows=1)
+    num_cols = first.shape[1]
+    if num_cols < 3:
+        raise ValueError(f"Expected at least 3 columns in BED file, got {num_cols}")
+    if num_cols > 12:
+        raise ValueError(f"Expected at most 12 columns in BED file, got {num_cols}")
+    if num_cols in (10, 11):
+        raise ValueError(f"BED10 and BED11 are prohibited, got {num_cols} columns")
+    if num_cols in (9, 12):
+        raise ValueError("BED9 and BED12 are valid but currently unsupported formats")
+    return BedType(num_cols)
+
+
+BZW_METADATA_FORMAT_VERSION = "0.1"
+
+
+@dataclasses.dataclass
+class BedZarrWriterMetadata(core.JsonDataclass):
```

Comment:
I wonder if we need this level of complexity for BED files. I had to work quite hard with VCF because things get arbitrarily big and it becomes necessary to allow for incomplete bits of fail and be restrarted. But, I feel like BED files are much smaller and could just be converted from pandas to zarr in a single shot, without worrying too much about optimal field sizes etc in the first instance.

What do you think @percyfal - is the layer of paranioa helpful here, or could be just get rid of it and be comfortable that a BED->conversion that fails could just be rerun entirely?

Reply:
@jeromekelleher I have removed large chunks of the code base. The current implementation relies on xarray.DataSet.to_zarr to do the actual conversion. I did keep the field type definitions to set the dtypes, but that could be removed too I guess.

Reply:
Ah, sorry @percyfal I should have been clear about this. Xarray is not a dependency we want in bio2zarr - it pulls in way too much stuff and depends on Dask. The motivation for writing this package in the first place was to move away from Xarray/Dask based methods for generating Zarrs! It's just not worth the extra complexity.

Reply:
I was imagining something simple like
```
array = zarr.empty(...)
array[:] = df[field]
```
So, you just set the zarr array in one go, letting Zarr deal with all the low-level details. It'll work fine at this scale.

Reply:
Ok, got it. Just push working code without xarray, but I'm guessing you want to go simpler still? I'm still picking up on Zarr as it were.

Reply:
This looks good, I think it's the right starting point. Let's get it cleaned up and merged :+1: